rtc_udh/src/lib.rs

```diff
@@ -0,0 +1,104 @@
+mod responder;
+
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex, RwLock},
+};
+
+use once_cell::sync::OnceCell;
+use responder::Responder;
+pub use responder::ResponderSys;
+use rtc_types::{
+    dh::{ExchangeReportResult, SessionRequestResult},
+    EcallResult,
+};
+use sgx_types::*;
+
+type SyncSendResponder = Arc<Mutex<Responder>>;
+
+type DhResponders = HashMap<sgx_enclave_id_t, SyncSendResponder>;
+
+fn dh_responders() -> &'static RwLock<DhResponders> {
+    static DH_RESPONDERS: OnceCell<RwLock<DhResponders>> = OnceCell::new();
+    DH_RESPONDERS.get_or_init(|| RwLock::new(HashMap::new()))
+}
+
+pub fn set_responder(
+    enclave_id: u64,
+    responder: Box<(dyn ResponderSys + 'static)>,
+) -> Result<(), sgx_status_t> {
+    match dh_responders().write() {
+        Ok(mut resp_map) => {
+            resp_map.insert(
+                enclave_id,
+                Arc::new(Mutex::new(Responder::new(enclave_id, responder))),
+            );
+            Ok(())
+        }
+        Err(_) => Err(sgx_status_t::SGX_ERROR_UNEXPECTED),
+    }
+}
+
+fn get_responder(id: &u64) -> Result<SyncSendResponder, sgx_status_t> {
+    dh_responders()
+        .read()
+        .or(Err(sgx_status_t::SGX_ERROR_UNEXPECTED))?
+        .get(id)
```

Comment:
```suggestion
        .get(&enclave_id)
```

(Following parameter.)